### PR TITLE
Import existing rclone.conf from non-portable locations

### LINKED ISSUE #2919: rclone: persist break the support of the existing configuration
Hi Guys, 
Sorry I'm bit late but this change break the support of the existing configuration as describe here : https://rclone.org/docs/#config-config-file

```
Rclone will look for an existing configuration file in any of the following locations, in priority order:

1. rclone.conf (in program directory, where rclone executable is)
2. %APPDATA%/rclone/rclone.conf (only on Windows)
3. $XDG_CONFIG_HOME/rclone/rclone.conf (on all systems, including Windows)
4. ~/.config/rclone/rclone.conf (see below for explanation of ~ symbol)
5. ~/.rclone.conf
```
Indeed, as the rclone.conf persist now, only the first options is considered by default/priority

My last update has broke my existing configuration, now I need to pass the --config flags each time....

_Originally posted by @PeeZu in https://github.com/ScoopInstaller/Main/issues/1634#issuecomment-964152900_

### PR DESCRIPTION
**Pull request by CrendKing**
Closes #2919
@rashil2000 

The import logic happens only when the `rclone.conf` file from persist does not exist or being empty. This check is here so that subsequent updates will not repeatedly import `rclone.conf` from non-portable locations and overwrite.

The import logic works like this:

Create an empty file in `$persist_dir` if the persist dir doesn't exist or the dir exists but no file in it (e.g. new installation). This basically forces a creation of the persist dir.

Then go through the location 2-5 in [this list](https://rclone.org/docs/#config-config-file), and test if a `rclone.conf` exists in any of those location. If yes, stop traversing and return the directory path as `$importConfDir`.

Finally, copy the config file from `$importConfDir` and print message about this event.

Because we have "persist" section, Scoop will link that imported file to app dir.

Also improved uninstall script so that cases that would result in error will now be avoided, such as app dir `rclone.conf` doesn't exist, or persist got deleted. Also, only copy if the link is broken, otherwise PowerShell complains that the source and target is the same file.

Do you think the logic is correct? Am I missing some corner cases?

### CODE REVIEW

**Review by rashil2000 — CHANGES_REQUESTED** [review]

File: bucket/rclone.json [hunk 1]
```diff
@@ -15,9 +15,28 @@
...
+        "   }",
+        "   if (Test-Path \"$importConfDir\\rclone.conf\") {",
+        "       Copy-Item \"$importConfDir\\rclone.conf\" $persist_dir",
+        "       Write-Host \"Imported existing rclone.conf from $importConfDir.\" -ForegroundColor Yellow",
+        "   }",
+        "}"
+    ],
     "uninstaller": {
-        "script": "Copy-Item \"$dir\\rclone.conf\" \"$persist_dir\\\" -Force"
+        "script": "if ((Test-Path \"$dir\\rclone.conf\") -and !(Get-Item \"$dir\\rclone.conf\").LinkType -and (Test-Path $persist_dir)) { Copy-Item \"$dir\\rclone.conf\" $persist_dir -Force }"
```
Comment (rashil2000) [inline bucket/rclone.json:39]:
```suggestion
        "script": "if ((Test-Path \"$dir\\rclone.conf\") -and !(Get-Item \"$dir\\rclone.conf\").LinkType -and (Test-Path $persist_dir)) { Copy-Item \"$dir\\rclone.conf\" $persist_dir }"
```

**Review by rashil2000 — CHANGES_REQUESTED** [review]

File: bucket/rclone.json [hunk 2]
```diff
@@ -15,9 +15,28 @@
...
+        "   if (Test-Path \"$env:APPDATA\\rclone\\rclone.conf\") {",
+        "       $importConfDir = \"$env:APPDATA\\rclone\"",
+        "   } elseif (Test-Path \"$env:XDG_CONFIG_HOME\\rclone\\rclone.conf\") {",
+        "       $importConfDir = \"$env:XDG_CONFIG_HOME\\rclone\"",
+        "   } else {",
+        "       if ($env:HOME) { $tilde = $env:HOME }",
+        "       elseif ($env:USERPROFILE) { $tilde = $env:USERPROFILE }",
+        "       else { $tilde = \"$env:HOMEDRIVE%\\$env:HOMEPATH\" }",
+        "       if (Test-Path \"$tilde\\.config\\rclone\\rclone.conf\") { $importConfDir = \"$tilde\\.config\\rclone\" }",
+        "       if (Test-Path \"$tilde\\rclone.conf\") { $importConfDir = $tilde }",
```
[the remark this answers is not in the document]
Comment (CrendKing) [reply bucket/rclone.json:30]: Nice catch!

File: bucket/rclone.json [hunk 3]
```diff
@@ -15,9 +15,28 @@
...
+        "if (!(Test-Path \"$persist_dir\\rclone.conf\") -or !(Get-Item \"$persist_dir\\rclone.conf\").Length) {",
+        "   New-Item \"$persist_dir\\rclone.conf\" -ItemType File -Force | Out-Null",
+        "   if (Test-Path \"$env:APPDATA\\rclone\\rclone.conf\") {",
+        "       $importConfDir = \"$env:APPDATA\\rclone\"",
+        "   } elseif (Test-Path \"$env:XDG_CONFIG_HOME\\rclone\\rclone.conf\") {",
+        "       $importConfDir = \"$env:XDG_CONFIG_HOME\\rclone\"",
+        "   } else {",
+        "       if ($env:HOME) { $tilde = $env:HOME }",
+        "       elseif ($env:USERPROFILE) { $tilde = $env:USERPROFILE }",
+        "       else { $tilde = \"$env:HOMEDRIVE%\\$env:HOMEPATH\" }",
```
[the remark this answers is not in the document]
Comment (CrendKing) [reply bucket/rclone.json:28]: Do you mean to use ternary operator to avoid the if blocks? Or you mean hoist just `$tilde` out and keep ifs? I'm not quite getting the "2nd nesting" part.

**Comment by CrendKing** [comment]
> Also, enclose $persist_dir in double quotes in all places, to account for paths with spaces.

When passing just the variable to a command, I'm pretty sure it doesn't need quotes. Like `Copy-Item 'foobar' $targetDir` and `Copy-Item 'foobar' "$targetDir"` are always equivalent, regardless of $targetDir's value.

File: bucket/rclone.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (rashil2000) [reply bucket/rclone.json]: Yes, declare $tilde outside, so that you don't have to write a nested if block.

[the remark this answers is not in the document]
Comment (CrendKing) [reply bucket/rclone.json]: Can you show me how to do it?

[the remark this answers is not in the document]
Comment (rashil2000) [reply bucket/rclone.json]: See 08ce9faff1f4c41fa65dbbda0878a862754ae453

[the remark this answers is not in the document]
Comment (CrendKing) [reply bucket/rclone.json]: 👍

**Comment by CrendKing** [comment]
Add a check for $importConf being $null, when there's nothing to import. Test-Path can't take $null argument.

**Comment by rashil2000** [comment]
How has the script been tested?

**Comment by CrendKing** [comment]
1. Fresh install, no persist, no import -> Empty config created
2. Update on 1 -> Nothing changed
3. Break link and fill some data to the app dir config, then update -> Config remained in app dir and copied to persist dir
4. Delete persist dir, then update -> Config is lost and reset to empty -> This is because we don't recreate persist dir during uninstaller script, but this should never happen anyways. Let me know if this should be changed
5. Prepare one rclone.conf for each location 2 through 5. Give each different content.
6. From 4, update, observe the "import" message and check file content -> Content matches the one from location 2
7. Delete config file from location 2, repeat last step, expect content change onward
8. Repeat until all are tested
9. To make sure we don't re-import, restore the config file from location 2, then update -> Observe there is no more importing

And I found a bug along the way 😉.

**Review by rashil2000 — APPROVED** [review]

**Comment by rashil2000** [comment]
Wow, thanks for the detailed report!

Looks good to me. 

Pinging @peezu and @erisa to confirm.

**Review by issaclin32 — APPROVED** [review]

**Comment by 7e4** [comment]
The persist logic works fine, until you edit the config file with 'rclone config'. It seems this is a bug of rclone not preserving hardlinks.

# pwd
C:\Users\user\scoop\apps\rclone\1.61.1

# fsutil hardlink list rclone.conf
\Users\user\scoop\persist\rclone\rclone.conf
\Users\user\scoop\apps\rclone\1.61.1\rclone.conf

# rclone config
After editing, hardlink disappears

# fsutil hardlink list rclone.conf
\Users\user\scoop\apps\rclone\1.61.1\rclone.conf

So it might be better to remove the persist logic and use defaults of rclone instead.

**Comment by rashil2000** [comment]
> It seems this is a bug of rclone not preserving hardlinks.

Has this issue been reported upstream? It would be good to get the developers' perspective on this.

**Comment by aphyra** [comment]
> 

it could be by design. I looked at the rclone pages first. It seems (on my pc) it renames it to rclone.conf.old and then writes a new one after finishing rclone config.
